File: Data Structures & Algorithms/palindrome-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        # find mid point
        # reverse second half
        # check if they are equal

        slow = head
        fast = head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        prev = None
        while slow:
            temp = slow.next
            slow.next = prev
            prev = slow
            slow = temp
        
        first = head
        second = prev
        while first:
            if first.val != second.val:
                return False
            first = first.next
            second = second.next
        return True
   
      

        
        """

        1 -> 2 -> 3 -> 4 -> Null

        1 -> 2 -> 3 -> Null

        """
        

File: Data Structures & Algorithms/valid-palindrome-ii/submission-1.cpp
class Solution {
public:
    bool validPalindrome(string s) {
        int l = 0;
        int r = s.size() - 1;
        while (l < r) {
            if (s[l] != s[r]) {
                return is_palindrome(s, l + 1, r) || is_palindrome(s, l , r - 1);
            }
            l += 1;
            r -= 1;
        }
        return true;
        
    }

    bool is_palindrome(const string& s, int l, int r) {
        while (l < r) {
            if (s[l] != s[r]) {
                return false;
            }
            l += 1;
            r -= 1;
        }
        return true;
    }
};